### PR TITLE
PT-159405994 get channel contract state

### PR DESCRIPTION
**Pull request by cytadela8**
PT [SC WS: Get contract and its state](https://www.pivotaltracker.com/story/show/159405994)

documentation PR: aeternity/protocol#321

### CODE REVIEW

**Comment by cytadela8** [comment]
Please check if tests are good enough. I removed assertion on `vm_version`, becouse for some reason locally I have `3` and on CI I get `1`.

File: apps/aechannel/src/aesc_offchain_state.erl [hunk 1]
```diff
@@ -288,6 +289,10 @@ get_latest_signed_tx(#state{signed_tx = LastSignedTx})
     when LastSignedTx =/= ?NO_TX ->
     {tx_round(aetx_sign:tx(LastSignedTx)), LastSignedTx}.
 
+-spec get_latest_signed_trees(state()) -> aec_trees:trees().
```
Comment (velzevur) [inline apps/aechannel/src/aesc_offchain_state.erl]: Why *signed* trees? Maybe `get_latest_trees/1`?
Cosmetical, so not a blocker

File: apps/aehttp/src/sc_ws_api_jsonrpc.erl [hunk 2]
```diff
@@ -349,6 +349,24 @@ process_request(#{<<"method">> := <<"channels.get.offchain_state">>,
...
                     , payload => aesc_offchain_state:serialize_for_client(State) }}
     end;
+process_request(#{<<"method">> := <<"channels.get.contract">>,
+                  <<"params">> := #{<<"pubkey">> := PubkeyE}}, FsmPid) ->
+    case aehttp_api_encoder:safe_decode(contract_pubkey, PubkeyE) of
+        {ok, Pubkey} ->
+            case aesc_fsm:get_contract(FsmPid, Pubkey) of
+                {ok, Contract} ->
+                    ContractState = aect_contracts:state(Contract),
+                    Resp = #{<<"contract">> => aect_contracts:serialize_for_client(Contract),
```
Comment (velzevur) [inline apps/aehttp/src/sc_ws_api_jsonrpc.erl]: I see the contract's serialization [does not include](https://github.com/aeternity/aeternity/blob/master/apps/aecontract/src/aect_contracts.erl#L204_L220) the state, hence the following line. Ok

File: apps/aechannel/src/aesc_offchain_state.erl (hunk 1 above)
Comment (cytadela8) [reply apps/aechannel/src/aesc_offchain_state.erl]: I added "signed" to signify it corresponds to the "signed_tx". To be honest it's a little bit confusing. I will change it.

File: apps/aehttp/src/sc_ws_api_jsonrpc.erl (hunk 2 above)
Comment (cytadela8) [reply apps/aehttp/src/sc_ws_api_jsonrpc.erl]: Yes, this makes sense as the contract state might be big. Here we specifically want to have it, but in general one might want only other parts of channel. Also it's stored separately.

File: apps/aechannel/src/aesc_offchain_state.erl [hunk 3]
```diff
@@ -27,6 +27,7 @@
         , set_half_signed_tx/2        %%  (SignedTx, State0) -> State
         , get_latest_half_signed_tx/1 %%  (State) -> SignedTx
         , get_latest_signed_tx/1      %%  (State) -> {Round, SignedTx}
+        , get_latest_trees/1   %%  (State) -> {Trees}
```
Comment (grzegorz225) [inline apps/aechannel/src/aesc_offchain_state.erl]: The comment stands out among the lines - add spaces before the comment.

File: apps/aehttp/src/aehttp_api_encoder.erl [hunk 4]
```diff
@@ -168,6 +168,8 @@ type2enc(block_state_hash) -> ?BASE58;
 type2enc(channel)          -> ?BASE58;
 type2enc(contract_pubkey)  -> ?BASE58;
 type2enc(contract_bytearray)-> ?BASE64;
+type2enc(contract_store_key)-> ?BASE64;
```
Comment (grzegorz225) [inline apps/aehttp/src/aehttp_api_encoder.erl]: Adjust spaces so the lines don't stand out.

File: apps/aehttp/src/aehttp_api_encoder.erl [hunk 5]
```diff
@@ -193,6 +195,8 @@ type2pfx(block_state_hash) -> <<"bs">>;
 type2pfx(channel)          -> <<"ch">>;
 type2pfx(contract_pubkey)  -> <<"ct">>;
 type2pfx(contract_bytearray)-> <<"cb">>;
+type2pfx(contract_store_key)-> <<"ck">>;
+type2pfx(contract_store_value)-> <<"cv">>;
```
Comment (grzegorz225) [inline apps/aehttp/src/aehttp_api_encoder.erl]: Same as above

File: apps/aehttp/src/aehttp_api_encoder.erl [hunk 6]
```diff
@@ -243,6 +249,8 @@ byte_size_for_type(block_state_hash) -> 32;
 byte_size_for_type(channel)          -> 32;
 byte_size_for_type(contract_pubkey)  -> 32;
 byte_size_for_type(contract_bytearray)-> not_applicable;
+byte_size_for_type(contract_store_key)-> not_applicable;
+byte_size_for_type(contract_store_value)-> not_applicable;
```
Comment (grzegorz225) [inline apps/aehttp/src/aehttp_api_encoder.erl]: Same as above

File: apps/aehttp/src/aehttp_api_encoder.erl (hunk 4 above)
Comment (cytadela8) [reply apps/aehttp/src/aehttp_api_encoder.erl]: I was done the same as existing code.

File: apps/aehttp/src/aehttp_api_encoder.erl (hunk 5 above)
Comment (cytadela8) [reply apps/aehttp/src/aehttp_api_encoder.erl]: I was done the same as existing code.

File: apps/aehttp/src/aehttp_api_encoder.erl (hunk 6 above)
Comment (cytadela8) [reply apps/aehttp/src/aehttp_api_encoder.erl]: I was done the same as existing code.

File: apps/aehttp/test/aehttp_integration_SUITE.erl [hunk 7]
```diff
@@ -3845,6 +3845,55 @@ query_state_(ConnPid, <<"json-rpc">>) ->
...
+
+sc_ws_get_contract(ConnPid, Pubkey, Config) ->
+    PubkeyE = aehttp_api_encoder:encode(contract_pubkey, Pubkey),
+    {ok, #{<<"contract">> := ContractE,
+           <<"contract_state">> := ContractStateE }} = query_contract(ConnPid, PubkeyE, Config),
+    #{ <<"id">>          := IdE
+    , <<"owner_id">>     := OwnerIdE
+    , <<"vm_version">>   := VMVersion
+    , <<"abi_version">>  := CTVersion
+    , <<"log">>          := Log
```
Comment (lucafavatella) [inline apps/aehttp/test/aehttp_integration_SUITE.erl]: I am checking whether exposing this field log makes sense. See also https://github.com/aeternity/aeternity/pull/2201

Comment (lucafavatella) [reply apps/aehttp/test/aehttp_integration_SUITE.erl]: This field here will always be empty binary, I suggest at least removing this field from this API as misleading. ([More detailed description](https://www.pivotaltracker.com/story/show/164539296).) Does user of this API - or other message - expect logs? If so, I expect this PR needs to extract those from calls.

Comment (cytadela8) [reply apps/aehttp/test/aehttp_integration_SUITE.erl]: Ok, I didn't have understanding of this field when making the PR. I think we will remove it from API for now and add it correctly if there is a need for it.

File: apps/aechannel/src/aesc_offchain_state.erl [hunk 8]
```diff
@@ -27,6 +27,7 @@
         , set_half_signed_tx/2        %%  (SignedTx, State0) -> State
         , get_latest_half_signed_tx/1 %%  (State) -> SignedTx
         , get_latest_signed_tx/1      %%  (State) -> {Round, SignedTx}
+        , get_latest_trees/1          %%  (State) -> {Trees}
```
Comment (lucafavatella) [inline apps/aechannel/src/aesc_offchain_state.erl]:
```suggestion
        , get_latest_trees/1          %%  (State) -> Trees
```

File: apps/aecontract/src/aect_contracts.erl [hunk 9]
```diff
@@ -205,15 +205,13 @@ serialize(#contract{owner_id     = OwnerId,
...
                                ct_version   = CTVersion,
-                               log          = Log,
                                active       = Active,
                                referrer_ids = ReferrerIds,
                                deposit      = Deposit}) ->
     #{ <<"id">>           => aehttp_api_encoder:encode(id_hash, Id)
      , <<"owner_id">>     => aehttp_api_encoder:encode(id_hash, OwnerId)
      , <<"vm_version">>   => maps:get(vm, CTVersion)
      , <<"abi_version">>  => maps:get(abi, CTVersion)
-     , <<"log">>          => Log
```
Comment (lucafavatella) [inline apps/aecontract/src/aect_contracts.erl]: I see this function used also in HTTP API, that declares this field as [required](https://github.com/aeternity/aeternity/blob/f5c46b80561697c2c41d03f9cc2dc7400b153e9e/config/swagger.yaml#L3102). Removing this from HTTP API seems backward incompatibility cc @noandrea - even if little - and we may postpone it ([tracked](https://www.pivotaltracker.com/story/show/164539296) anyway). Though anyway if we change it in code we need to update swagger schema.

May I suggest in this PR to keep user HTTP API unchanged, and just not expose this fild in websocket? (I understand not cleanest in code though least impact on API.)

File: apps/aehttp/src/aehttp_api_encoder.erl [hunk 10]
```diff
@@ -160,54 +160,58 @@ type2id(contract_pubkey) -> contract;
...
-type2enc(call_state_tree)  -> ?BASE64.
+type2enc(key_block_hash)       -> ?BASE58;
+type2enc(micro_block_hash)     -> ?BASE58;
+type2enc(block_pof_hash)       -> ?BASE58;
+type2enc(block_tx_hash)        -> ?BASE58;
+type2enc(block_state_hash)     -> ?BASE58;
+type2enc(channel)              -> ?BASE58;
+type2enc(contract_pubkey)      -> ?BASE58;
+type2enc(contract_bytearray)   -> ?BASE64;
+type2enc(contract_store_key)   -> ?BASE64;
```
Comment (lucafavatella) [inline apps/aehttp/src/aehttp_api_encoder.erl]: May I kindly ask you refresh [the tests in `aehttp_api_encoder_tests` for the contract-related encodings](https://github.com/aeternity/aeternity/blob/f5c46b80561697c2c41d03f9cc2dc7400b153e9e/apps/aehttp/test/aehttp_api_encoder_tests.erl#L10-L26)?

**Comment by cytadela8** [comment]
@lucafavatella removal of the log field is the reason for test failure. Should I maybe re-add the field so it will be returned by both normal contract related API and get.channel.contract_state method?

File: apps/aecontract/src/aect_contracts.erl (hunk 9 above)
Comment (cytadela8) [reply apps/aecontract/src/aect_contracts.erl]: Ok, I will do that!

**Comment by lucafavatella** [comment]
I took the liberty of labelling as WIP. Requires rebase.

File: apps/aehttp/test/aehttp_integration_SUITE.erl [hunk 11]
```diff
@@ -4259,13 +4333,25 @@ sc_ws_oracle_contract_(Owner, GetVolley, ConnPid1, ConnPid2,
...
                   <<"string">>, <<"ok">>),
+    #{ active := true
+     , ct_version := 1
+     , deposit := 10
+     , id := ContractId
+     , owner_id := OwnerId
+     , referrer_ids := []
+     , state := State4
+    } = sc_ws_get_contract_assert_equal(ConnPid1, ConnPid2, ContractPubKey, Config),
+    <<0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,4>> = maps:get(<<2>>, State4),
```
Comment (lucafavatella) [inline apps/aehttp/test/aehttp_integration_SUITE.erl]: A comment on these `<<2>>` and related values ease readability of the test. Though debatable so ok.

**Comment by cytadela8** [comment]
@lucafavatella I added a relase-note

**Comment by cytadela8** [comment]
I have adjusted the release note partially to trigger AppVeyour CI.

**Pull request merged**